lib/rate-limit.js

```diff
@@ -1,86 +1,86 @@
 import { Promise } from 'es6-promise'
 
-import promisedWait from './promised-wait'
-
 /**
- * Promise-based rate limiting of a function. Assumes that the passed in
- * function will return a Promise.
- * Queues calls made to the wrapped function, and if more calls are made than
- * the allowed level of concurrency waits until a certain amount of time has
- * passed before continuing to clear calls from the queue.
+ * Promise-based rate limiting for a given set of methods of one object.
+ * Assumes that the limited functions will return a Promise.
+ * Calls made to any wrapped function will be stored in a single queue.
+ * If more calls are added at once than the allowed level of concurrency, it
+ * waits until a certain amount of time has passed before continuing to clear
+ * calls from the queue.
  * @private
- * @param {Function} fn - call to be rate limited
+ * @param {object} client - client object whose methods should be limited
+ * @param {array} methodsToLimit - Array of methods which should be limited
  * @param {number} concurrency - Number of allowed concurrent requests
  * @param {number} delay - Delay in milliseconds for waiting after hitting the
  */
-export default function rateLimit (fn, concurrency, delay) {
-  concurrency = positiveInteger('concurrency', concurrency)
-  delay = positiveInteger('delay', delay)
-  const callQueue = []
-  let inFlight = 0
+export default function rateLimit (client, methodsToLimit, concurrency, delay) {
+  // Wrapper function to limit given methods.
+  const enqueue = (method) => {
+    return () => {
+      return new Promise((resolve) => {
+        queue.push({
+          args: Array.prototype.slice.call(arguments),
```

Comment:
To which `arguments` are you refering here? Arrow functions don't have an own`arguments` object and will refer to the first available one on the scope (the one for `rateLimit` on this case).

It could be possible that on this case, since the code is transpiled everything works magically because the arrow function is converted to `function` :man_shrugging:.

In any case, using the `arguments` object like that is considered a performance killer https://github.com/vhf/v8-bailout-reasons#bad-value-context-for-arguments-value so we might consider not using them.